chef-config/lib/chef-config/path_helper.rb

```diff
@@ -55,18 +55,24 @@ def self.path_separator
       end
     end
 
-    def self.join(*args)
-      path_separator_regex = Regexp.escape(File::SEPARATOR)
-      unless path_separator == File::SEPARATOR
-        path_separator_regex << Regexp.escape(path_separator)
-      end
+    def self.path_separator_regex
+      @path_separator_regex ||=
+        begin
+          path_separator_regex = Regexp.escape(File::SEPARATOR)
+          unless path_separator == File::SEPARATOR
+            path_separator_regex << Regexp.escape(path_separator)
+          end
+          path_separator_regex
+        end
+    end
 
-      trailing_slashes = /[#{path_separator_regex}]+$/
-      leading_slashes = /^[#{path_separator_regex}]+/
+    TRAILING_SLASHES_REGEX = /[#{path_separator_regex}]+$/.freeze
+    LEADING_SLASHES_REGEX = /^[#{path_separator_regex}]+/.freeze
 
+    def self.join(*args)
       args.flatten.inject do |joined_path, component|
-        joined_path = joined_path.sub(trailing_slashes, "")
-        component = component.sub(leading_slashes, "")
+        joined_path = joined_path.sub(TRAILING_SLASHES_REGEX, "")
+        component = component.sub(LEADING_SLASHES_REGEX, "")
         joined_path + "#{path_separator}#{component}"
       end
```

Comment:
This is still likely doing more allocations than necessary because of building strings with inject and multiple `.sub` calls. Something like this might do fewer:

```
irb(main):036:0> a = ["/foo/", "bar/", "/baz", "////tacos//"]
=> ["/foo/", "bar/", "/baz", "////tacos//"]
irb(main):037:0> path_separator = File::SEPARATOR
=> "/"
irb(main):038:0> a.map {|s| s[/^[#{path_separator}]*(.*?)[#{path_separator}]*$/, 1] }.join(path_separator)
=> "foo/bar/baz/tacos"
```

Reply:
that's a little crazy to read though and we're down by something like 100x less allocations now

Reply:
Haha, understandable. 👍